pkg/control/control.go

```diff
@@ -204,59 +210,57 @@ var (
 
 func (ctrl *Control) setupHeartbeat(channelID types.ChannelID) {
 	ticker := time.NewTicker(15 * time.Second)
-	go func() {
-		tickFailed := 0
+	tickFailed := 0
 
-		stream, err := ctrl.getStream(channelID)
-		if err != nil {
-			return
-		}
+	stream, err := ctrl.getStream(channelID)
+	if err != nil {
+		return
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			stream.log.Infof("Collecting metadata tickFailed=%d", tickFailed)
+			var err error
+			hasErrors := false
+
+			err = ctrl.sendThumbnail(channelID)
+			if err != nil {
+				stream.log.Error(errors.Wrap(err, ErrHeartbeatThumbnail.Error()))
+				hasErrors = true
+			}
 
-		for {
-			select {
-			case <-ticker.C:
-				stream.log.Infof("Collecting metadata tickFailed=%d", tickFailed)
-				var err error
-				hasErrors := false
-
-				err = ctrl.sendThumbnail(channelID)
-				if err != nil {
-					stream.log.Error(errors.Wrap(err, ErrHeartbeatThumbnail.Error()))
-					hasErrors = true
-				}
-
-				err = ctrl.sendMetadata(channelID)
-				if err != nil {
-					stream.log.Error(errors.Wrap(err, ErrHeartbeatSendMetadata.Error()))
-					hasErrors = true
-				}
-
-				err = ctrl.orchestrator.Heartbeat(channelID)
-				if err != nil {
-					stream.log.Error(errors.Wrap(err, ErrHeartbeatOrchestratorHeartbeat.Error()))
-					hasErrors = true
-				}
-
-				if hasErrors {
-					tickFailed++
-				} else if tickFailed > 0 {
-					tickFailed--
-				}
-
-				// Look for 3 consecutive failures
-				if tickFailed >= 5 {
-					stream.log.Warn("Stopping stream due to excessive heartbeat errors")
-					ctrl.StopStream(channelID)
-					ticker.Stop()
-					return
-				}
-
-			case <-ctrl.metadataCollectors[channelID]:
+			err = ctrl.sendMetadata(channelID)
+			if err != nil {
+				stream.log.Error(errors.Wrap(err, ErrHeartbeatSendMetadata.Error()))
+				hasErrors = true
+			}
+
+			err = ctrl.orchestrator.Heartbeat(channelID)
+			if err != nil {
+				stream.log.Error(errors.Wrap(err, ErrHeartbeatOrchestratorHeartbeat.Error()))
+				hasErrors = true
+			}
+
+			if hasErrors {
+				tickFailed++
+			} else if tickFailed > 0 {
+				tickFailed--
+			}
+
+			// Look for 3 consecutive failures
+			if tickFailed >= 5 {
+				stream.log.Warn("Stopping stream due to excessive heartbeat errors")
+				stream.Stop()
```

Comment:
Actually, strike that, it bubbles up and FTL calls StopStream! There's room to cleanup these "Stop" names in the future lol

Reply:
I had it like you say first, however, that ended up in a double removal of the stream. This is because the `OnVideo`/`OnAudio` handlers also call the `StopStream` method.

The second invocation of `StopStream` only results in an early return with an error `stream does not exist in state` so it's harmless but it just shows up in the logs as an error.

Should I change it to `ctrl.StopStream(channelID)`?

Reply:
It's good how you have it in the PR! I dug in a little deeper after leaving the comment.